docs/guide/programming-tasks.md

```diff
@@ -0,0 +1,288 @@
+choose Starlark instead of common alternatives like JavaScript/TypeScript or WebAssembly:
+
+1. Starlark doesn't require compilation. No need to introduce full-blown compile -> deploy process for a few dozen lines of logic.
+2. Starlark script can be executed instantly on any platform. There is Starlark interpreter written in Go which integrates nicely with Cirrus CLI / Cirrus CI infrastructure.
+3. Starlark has built-in functionality for loading external modules which is ideal for config sharing. See [module loading](#module-loading) for details.
+
+## Writing Starlark scripts
+
+Let's start with a trivial `.cirrus.star` example like this:
+
+```python
+def main():
+    return [
+        {
+            "container": {
+                "image": "debian:latest",
+            },
+            "script": "make",
+        },
+    ]
+```
+
+`main` simply returns a list of task objects which will be serialized into YAML presentation like this:
+
+```yaml
+task:
+    container:
+      image: debian:latest
+    script: make
+```
+
+Then the generated YAML is appended to `.cirrus.yml` (if any) before passing the combined config into the final YAML parser.
+
+With Starlark, it's possible to generate parts of the configuration dynamically based on some external conditions: by [making an HTTP request](#http) to check the previous build status
+or by [parsing files inside the repository](#fs) to pick up some common settings (for example, parse `package.json` to see if it contains `lint` script and generate a linting task).
+
+Even more importantly: with the [module loading](#module-loading) you can re-use other people's code to avoid wasting time on things written from scratch.
+For example, there are official [task helpers](https://github.com/cirrus-modules/helpers) available that reduce the boilerplate when generating tasks:
+
+```python
+load("github.com/cirrus-modules/helpers", "task", "container", "script")
+
+def main(ctx):
+  return [
+    task(
+      instance=container("debian:latest"),
+      instructions=[script("make")]
+    ),
+  ]
+```
+
+### Entrypoints
+
+Different events will trigger execution of different top-level functions in the `.cirrus.star`. These functions reserve certain names
+and will be called with different arguments depending on the event which triggered the execution.
+
+#### `main`
+
+`main` is called once a Cirrus CI build is triggered in order to generate a list of tasks to execute within that particular build.
+
+#### Hooks
+
+It's also possible to execute Starlark scripts on updates to the current build or any of the tasks within the build.
+Think of it as [WebHooks](../api.md#webhooks) running within Cirrus that doesn't require any infrastructure on your end.
+
+Expected names of Starlark Hook functions in `.cirrus.star` are `on_build_<STATUS>` or `on_task_<STATUS>` respectively.
+These functions should expect a single context argument passed by Cirrus Cloud. At the moment hook's context only contains
```

Comment:
This looks like a regression from https://github.com/cirruslabs/cirrus-cli/blob/236173c844d56c923b8f0c932bb8a04550cd7e15/STARLARK.md#entrypoints to me.

In CLI's docs, a short glance below the entrypoint allows the user to discover arguments and a return value and skim across if needed.

Also, https://github.com/cirruslabs/cirrus-cli/issues/386 talks about `main()`, but do we want to mirror the same feature for hooks?

Reply:
I think it's a bit less documentation like in terms of explicitness. Right now there is only a single context is passed. In case there will options in future we can rework it similar to how it was.

And yeah, I created cirruslabs/cirrus-cli#386 to allow skip mentioning of the context here. Once we'll think of anything worthy to add to the context we can rework this part. Right now `ctx` is always empty and IMO should be omitted.